Feature/1.15/warehouse inventory, new inventory tab

lexustec opened the pull request:
![GUI_Proposal](https://user-images.githubusercontent.com/4784045/79476144-e64b2700-8008-11ea-98f1-34982f831955.png)
# Changes proposed in this pull request:
- creates a new tab in the Warehose screen which displays the Inventory of all Racks belonging to that Warehouse
- a lot of Formating changes due re run of the code formating settings in IntelliJ

Currently requires #4634 to be merged first

Review please

universemodifier committed:
Change in BuilderAi to be a little bit more smart---still has a bug

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real values.

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real values.

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15

universemodifier committed:
Change in BuilderAi to be a little bit more smart

Bugfix... endless loop at bigger buildings

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

universemodifier committed:
Modified and moved AI code

Changed the System to reflect more proper requests if needed material is bigger than stacksize. and also to pick not only full stacks.. instead real needed values.

universemodifier committed:
Merge remote-tracking branch 'origin/version/1.15' into version/1.15

universemodifier committed:
formatting change

fixed formatting

universemodifier committed:
Fix and format

universemodifier committed:
Format Fix

Fixed again. now should be good

universemodifier committed:
Tuple replace

Now it should be ready for review

universemodifier committed:
Warehouse Inventory Total page working

The Warehouse got a new page which lists now all Warehouse connected Inventory. Icons still missing and window text is abstract for now. I didnt want to fuzz in the Language files for now. correct name needs to be found. Tech wise works nice. Ready for test

someaddons commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingWareHouse.java:
```diff
@@ -525,6 +537,8 @@ else if (request != null)
          */
         private boolean reachedLimit = false;
 
+        private List<BlockPos> storageracks = new ArrayList<>();
```
our buildings already have a list of containers, called containerList. So you don't need to add all this stuff

lexustec replied:
This is in the deserialize which is accessed by the View. i can not access the containerList there.. i tried a lot to access the container list but failed.

someaddons replied:
oh its in the view, then you probably don't need it at all. The warehouse tileentity has a combined inventory of all rakcs

lexustec replied:
oO okay ? this i built separately in the View. it iterates through all racks and adds up.

someaddons replied:
ye you made more effort than you need to xD

someaddons replied:
sth like tileentity.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY,null) and you get an already combined IItemhandler

marchermans replied:
Yep this will need to be removed for simplicities sake.

Raycoms replied:
I think how it is right now it is more efficient since he gets the cached content list @OrionDevelopment

someaddons commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1435,58 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
      */
-    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Predicate<ItemStack> predicate)
+
+    private boolean tryTransferFromPosToWorker(final BlockPos pos, @NotNull final Tuple<Predicate<ItemStack>, Integer> predicate)
```
make it two parameters isntead of the tuple

lexustec replied:
The Itemlist created at  is a Tuple<Predicate<ItemStack>, Integer> because the needsCurrently var as the transfer var is feeded in the workers. and the Predicate<ItemStack> doesnt transfer the count.

someaddons commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:
```diff
@@ -1463,19 +1435,58 @@ public boolean checkIfRequestForItemExistOrCreateAsynch(@NotNull final ItemStack
+        }
+        else
+        {
+            return true; // has already needed transfers...
```
sure this is right? tryTransferFromPosToWorker doesnt seem like a function which checks the own inventory for contains

lexustec replied:
this is why i added getItemCountInItemHandler. which assists in picking the correct amount. like if i need 2 stacks he will pick 2 stacks ( the Do while loop ). prevents that a Builder walks multiple times for the same stuff.

someaddons replied:
ye then rename the function to sth like transferItemsIfNeeded or so

lexustec replied:
i can. is only used by pickupMaterials() anyway.

someaddons commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java:
```diff
@@ -164,19 +169,20 @@ private IAIState queryItems()
                 if (InventoryUtils.hasItemInProvider(getOwnBuilding(), predicate))
                 {
-                    needsCurrently = predicate;
+                    needsCurrently = new Tuple<>(predicate, Constants.STACKSIZE);
```
why Constants.STACKSIZE instead of inputStorage.amount?

lexustec replied:
because i experienced that ItemStack sometimes has more than STACKSIZE ( see in pickupMaterial() when it creates the list.)

universemodifier committed:
Reformed Itemcollection and Filters

Added features to the screen and fixed icon display.

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5366054&bid=17583621)

universemodifier committed:
Minor changes

changed icon display, added language entry, removed inventory button and resized list

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5366054&bid=17583621)

universemodifier committed:
Added code for big recipes to average out items

marvin-bitterlich commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java  4
- src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5366054&bid=17583621)

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -107,6 +146,37 @@ private void addStock()
         }
     }
 
+    void setSortFlag()
```
private + javadoc please

lexustec replied:
done

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -36,10 +46,36 @@
+    public static final int           COUNT_ASC_SORT  = 3;
+    public static final int           COUNT_DESC_SORT = 4;
+    int sortDescriptor = 0;
+
```
javadoc for the remaining options please

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutWareHouse.java:
```diff
@@ -247,6 +315,151 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
+
+        for (int s = 0; s < count; s++)
+        {
+            final TileEntity rack = world.getTileEntity(containerList.get(s));
```
Btw, there is already a way to get ItemStorage from the rack without having to go through all slots

lexustec replied:
I need to go through all slots of all Racks because i am summing all up. because Multiple Stacks of the same type can be in multiple Racks. This function creates basically a single ItemStack List with all combined.

The getContainerlist returns only the block position from the building. i couldnt access it directly because i am in the View part. so i need to serialize and deserialize it. i serialized the containerList from the building side and deserialized it in the view part. from there i get the inventory of each rack. and then i need to sum up all together. then adding the filter and set it to the allItems List and call the updateResourceList

universemodifier committed:
Formatting changes and Javadoc

universemodifier committed:
Changing in how to get the stacks to the view

universemodifier committed:
Merge remote-tracking branch 'upstream/version/1.15' into feature/1.15/warehouse_inventory

universemodifier committed:
Mergefix

universemodifier committed:
Updated Position of Inventory Stuff

Window in Inventoryview bugs left...

universemodifier committed:
small fixes

universemodifier committed:
Positioning fix

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java:
```diff
@@ -0,0 +1,251 @@
+        for (int s = 0; s < count; s++)
+        {
+            final TileEntity rack = world.getTileEntity(containerList.get(s));
+            if (rack instanceof TileEntityRack)
```
if not instanceof TileEntityRack but rack.getCapability(IItemHandlerCap) then  get itemhandler and iterate over slots.

universemodifier committed:
Added ChestTileEntity in Inventory slot check

universemodifier committed:
if to else if

marchermans reviewed (approved):
All in all very good work.
One minor nitpicky thing is that instead of casting to chests, i would rather that you asked our inventory utils to get you all stacks in the TE, but only if it is not a rack.

Raycoms committed:
small fixes (#4691)

Fix some university issues
Add some more block recipes to the stonemason

Raycoms committed:
Hotfix allow to detect non minecolonies hutblock

Raycoms committed:
fix imports

Raycoms committed:
hotfix max happiness (#4692)

Max happiness fix

Raycoms committed:
fix child sounds together, fix placement handling not using forgereg (#4693)

Declutters sounds a little bit (step 1)
Adds a better way for hut detection for placement handling

melv-n commented:
New player of the mod here, loving this PR! Not being able to find items or organise them was one of my main gripes with the warehouse. This is great!

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutAllInventory.java:
```diff
@@ -0,0 +1,268 @@
+    public boolean onKeyTyped(final char ch, final int key)
+    {
+        final boolean result = super.onKeyTyped(ch, key);
+        filter = findPaneOfTypeByID("names", TextField.class).getText();
```
if result == true (since else means the event was not caught by this gui)

universemodifier committed:
Merge remote-tracking branch 'upstream/version/1.15' into feature/1.15/warehouse_inventory
